include/knowhere/comp/rw_lock.h

```diff
@@ -0,0 +1,64 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+#ifndef KNOWHERE_RW_LOCK_H
+#define KNOWHERE_RW_LOCK_H
+#include <atomic>
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+/*
+FairRWLock is a fair MultiRead-SingleWrite lock
+*/
+namespace knowhere {
+class FairRWLock {
+ public:
+    void
+    LockRead() {
+        std::unique_lock lock(mtx_);
+        auto cur_id = task_id_counter_++;
+        task_id_q.push(cur_id);
+        task_cv_.wait(lock, [&] { return task_id_q.front() == cur_id && !have_writer_task_; });
+        reader_task_counter_++;
+        task_id_q.pop();
+    }
+    void
+    UnLockRead() {
+        std::unique_lock lock(mtx_);
+        if (--reader_task_counter_ == 0) {
+            task_cv_.notify_all();
+        }
+    }
```

Comment:
Please wrap this in RAII to avoid failures that would cause the lock to not be released